ngsPETSc/utils/firedrake/adaptive_hierarchy.py

```diff
@@ -0,0 +1,331 @@
+        self._shared_data_cache = defaultdict(dict)
+
+        if len(self.meshes) <= 2:
+            self.coarse_to_fine_cells = {}
+            self.fine_to_coarse_cells = {}
+            self.fine_to_coarse_cells[Fraction(0, 1)] = None
+
+        # extract parent child relationships from netgen meshes
+        (c2f, f2c) = get_c2f_f2c_fd(mesh, coarse_mesh)
+
+        self.coarse_to_fine_cells[Fraction(len(self.meshes) - 2, 1)] = c2f
+        self.fine_to_coarse_cells[Fraction(len(self.meshes) - 1, 1)] = np.array(f2c)
+
+        # split both the fine and coarse meshes into the submeshes
+        (coarse_splits, fine_splits, num_children) = split_to_submesh(
+            mesh, coarse_mesh, c2f, f2c
+        )
+        for i in range(1, max_children + 1):
+            coarse_mesh.mark_entities(coarse_splits[i], i)
+            mesh.mark_entities(fine_splits[i], int(f"10{i}"))
+
+        coarse_mesh = RelabeledMesh(
+            coarse_mesh,
+            [coarse_splits[i] for i in range(1, max_children + 1)],
+            list(range(1, max_children + 1)),
+            name="Relabeled_coarse",
+        )
+        c_subm = {
+            j: Submesh(coarse_mesh, coarse_mesh.topology_dm.getDimension(), j)
+            for j in range(1, max_children + 1)
+            if any(num_children == j)
+        }
+        set_level(coarse_mesh, self, level - 2)
+
+        mesh = RelabeledMesh(
+            mesh,
+            [fine_splits[i] for i in range(1, max_children + 1)],
+            list(range(1, max_children + 1)),
+        )
+        f_subm = {
+            int(str(j)[-2:]): Submesh(mesh, mesh.topology_dm.getDimension(), j)
+            for j in [int("10" + str(i)) for i in range(1, max_children + 1)]
+            if any(num_children == int(str(j)[-2:]))
+        }
+        set_level(mesh, self, level - 1)
+
+        # update c2f and f2c for submeshes by mapping numberings
+        # on full mesh to numberings on coarse mesh
+        n = [
+            len([el for el in c2f if len(el) == j]) for j in range(1, max_children + 1)
+        ]
+        c2f_adjusted = {
+            j: np.zeros((n, j))
+            for n, j in zip(n, list(range(1, max_children + 1)))
+            if n != 0
+        }
+        f2c_adjusted = {
+            j: np.zeros((n * j, 1))
+            for n, j in zip(n, list(range(1, max_children + 1)))
+            if n != 0
+        }
+
+        coarse_full_to_sub_map = {
+            i: full_to_sub(coarse_mesh, c_subm[i], i) for i in c_subm
+        }
+        fine_full_to_sub_map = {
+            j: full_to_sub(mesh, f_subm[j], int("10" + str(j))) for j in f_subm
+        }
+
+        for i, children in enumerate(c2f):
+            n = len(children)
+            if 1 <= n <= max_children:
+                c2f_adjusted[n][coarse_full_to_sub_map[n](i)] = fine_full_to_sub_map[n](
+                    np.array(children)
+                )
+
+        for j, parent in enumerate(f2c):
+            n = int(num_children[parent].item())
```

Comment:
Is this just to avoid numpy types? What happens if you do?
```suggestion
            n = num_children[parent]
```

Reply:
The following loop after this will return that n is an unhashable type: 'numpy.ndarray' , but you are right that there is no need for the int()